gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaAvroSchemaRegistry.java

```diff
@@ -0,0 +1,289 @@
+
+    Schema schema;
+    if (schemaString.startsWith("{")) {
+      try {
+        schema = new Schema.Parser().parse(schemaString);
+      } catch (Exception e) {
+        throw new SchemaNotFoundException(String.format("Latest schema for topic %s cannot be retrieved", topic));
+      }
+    } else {
+      throw new SchemaNotFoundException(String.format("Latest schema for topic %s cannot be retrieved", topic));
+    }
+
+    return schema;
+  }
+
+  /**
+   * Register a schema to the Kafka schema registry
+   *
+   * @param schema
+   * @return schema ID of the registered schema
+   */
+  public synchronized String register(Schema schema) {
+    LOG.info("Registering schema " + schema.toString());
+
+    PostMethod post = new PostMethod(url);
+    post.addParameter("schema", schema.toString());
+
+    int statusCode;
+    try {
+      LOG.debug("Loading: " + post.getURI());
+      statusCode = httpClient.executeMethod(post);
+      if (statusCode != HttpStatus.SC_CREATED)
+        LOG.error("Error occurred while trying to register schema: " + statusCode);
+    } catch (URIException e) {
+      throw new RuntimeException(e);
+    } catch (HttpException e) {
+      throw new RuntimeException(e);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    } finally {
+      post.releaseConnection();
+    }
+
+    LOG.info("Registered schema successfully");
+    String response;
+    try {
+      response = post.getResponseBodyAsString();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+
+    String schemaId;
+    Header[] headers = post.getResponseHeaders(SCHEMA_ID_HEADER_NAME);
+    if (headers.length != 1) {
+      LOG.error("Error reading schemaId returned by registerSchema call");
+      throw new RuntimeException("Error reading schemaId returned by registerSchema call");
+    } else if (!headers[0].getValue().startsWith(SCHEMA_ID_HEADER_PREFIX)) {
+      LOG.error("Error parsing schemaId returned by registerSchema call");
+      throw new RuntimeException("Error parsing schemaId returned by registerSchema call");
+    } else {
+      schemaId = headers[0].getValue().substring(SCHEMA_ID_HEADER_PREFIX.length());
+    }
+
+    if (response != null) {
+      LOG.info("Received response " + response);
+    }
+
+    return schemaId;
+  }
+
+  private class KafkaSchemaCacheLoader extends CacheLoader<String, Schema> {
+
+    private final Map<String, FailedFetchHistory> failedFetchHistories;
+
+    private KafkaSchemaCacheLoader() {
+      super();
+      this.failedFetchHistories = Maps.newHashMap();
+    }
+
+    @Override
+    public Schema load(String id) throws Exception {
+      if (shouldFetchFromSchemaRegistry(id)) {
+        try {
+          return fetchSchemaByID(id);
+        } catch (SchemaNotFoundException e) {
+          addFetchToFailureHistory(id);
+          throw e;
+        }
+      }
+      throw new SchemaNotFoundException(String.format("Schema with ID = %s cannot be retrieved", id));
```

Comment:
So an exception is thrown if `shouldFetchFromSchemaRegistry` returns false. Can you explain why doing this and add comment in the code?